File: internal/app/dto/allocation_policy.go
package dto

import (
	"encoding/json"
	"time"
)

// ---- AllocationPolicyRule CRUD ----

type AllocationPolicyRuleDTO struct {
	ID                   uint            `json:"id"`
	WaveID               uint            `json:"waveId"`
	ProductID            uint            `json:"productId"`
	SelectorPayload      json.RawMessage `json:"selectorPayload"`
	ProductTargetRef     string          `json:"productTargetRef"`
	ContributionQuantity int             `json:"contributionQuantity"`
	RuleKind             string          `json:"ruleKind"`
	Priority             int             `json:"priority"`
	Active               bool            `json:"active"`
	CreatedAt            time.Time       `json:"createdAt" ts_type:"string"`
	UpdatedAt            time.Time       `json:"updatedAt" ts_type:"string"`
}

type CreateAllocationPolicyRuleInput struct {
	WaveID               uint            `json:"waveId"`
	ProductID            uint            `json:"productId"`
	SelectorPayload      json.RawMessage `json:"selectorPayload"`
	ProductTargetRef     string          `json:"productTargetRef"`
	ContributionQuantity int             `json:"contributionQuantity"`
	RuleKind             string          `json:"ruleKind"`
	Priority             int             `json:"priority"`
	Active               bool            `json:"active"`
}

type UpdateAllocationPolicyRuleInput struct {
	ID                   uint             `json:"id"`
	ProductID            *uint            `json:"productId,omitempty"`
	SelectorPayload      *json.RawMessage `json:"selectorPayload,omitempty"`
	ProductTargetRef     *string          `json:"productTargetRef,omitempty"`
	ContributionQuantity *int             `json:"contributionQuantity,omitempty"`
	RuleKind             *string          `json:"ruleKind,omitempty"`
	Priority             *int             `json:"priority,omitempty"`
	Active               *bool            `json:"active,omitempty"`
}

// ---- Reconcile result ----

type ReconcileResultDTO struct {
	Created       int                `json:"created"`
	Deleted       int                `json:"deleted"`
	ReplayedCount int                `json:"replayedCount"`
	Failures      []ReplayFailureDTO `json:"failures"`
}

type ReplayFailureDTO struct {
	AdjustmentID uint   `json:"adjustmentId"`
	Reason       string `json:"reason"`
}

File: frontend-next/src/entities/allocation-policy.ts
/**
 * AllocationPolicy entity types.
 *
 * Unlike other entity files (which re-export generated DTO shapes), this file
 * defines its own interfaces because the Wails codegen represents
 * `selector_payload` as `number[]`, whereas the actual runtime value is a
 * typed discriminated union (`SelectorPayload`). The entity types here are
 * the canonical frontend representation; callers MUST use these instead of
 * the generated `dto.AllocationPolicyRuleDTO` / `dto.CreateAllocationPolicyRuleInput` / etc.
 */

/** Selector payload — discriminated union by `type` field. */
export interface SelectorPayload {
  type: "wave_all" | "platform_all" | "identity_level" | "explicit_override"
  platform?: string
  level?: string
  participant_ids?: number[]
}

/**
 * AllocationPolicyRule (aligned to Go dto.AllocationPolicyRuleDTO).
 * Container fields use camelCase to match the DTO json tags; the nested
 * `selectorPayload` keys (`type`/`platform`/`level`/`participant_ids`) stay
 * as-is because they mirror the domain SelectorPayload JSON stored in the DB.
 */
export interface AllocationPolicyRule {
  id: number
  waveId: number
  productId: number
  selectorPayload: SelectorPayload
  productTargetRef: string
  contributionQuantity: number
  ruleKind: string
  priority: number
  active: boolean
  createdAt: string
  updatedAt: string
}

/** Input for creating a new allocation policy rule. */
export interface CreateAllocationPolicyRuleInput {
  waveId: number
  productId: number
  selectorPayload: SelectorPayload
  productTargetRef: string
  contributionQuantity: number
  ruleKind: string
  priority: number
  active: boolean
}

/** Input for updating an existing allocation policy rule. Partial fields. */
export interface UpdateAllocationPolicyRuleInput {
  id: number
  productId?: number
  selectorPayload?: SelectorPayload
  productTargetRef?: string
  contributionQuantity?: number
  ruleKind?: string
  priority?: number
  active?: boolean
}

/** Result of a reconcile operation. */
export interface ReconcileResult {
  created: number
  deleted: number
  replayedCount: number
  failures: ReplayFailure[]
}

/** A single replay failure entry. */
export interface ReplayFailure {
  adjustmentId: number
  reason: string
}
